openmc/geometry.py

```diff
@@ -0,0 +1,610 @@
+
+        """
+        material_cells = OrderedDict()
+
+        for cell in self.get_all_cells().values():
+            if cell.fill_type in ('material', 'distribmat'):
+                if cell not in material_cells:
+                    material_cells[cell.id] = cell
+
+        return material_cells
+
+    def get_all_material_universes(self):
+        """Return all universes having at least one material-filled cell.
+
+        This method can be used to find universes that have at least one cell
+        that is filled with a material or is void.
+
+        Returns
+        -------
+        collections.OrderedDict
+            Dictionary mapping universe IDs to :class:`openmc.Universe`
+            instances with at least one material-filled cell
+
+        """
+        material_universes = OrderedDict()
+
+        for universe in self.get_all_universes().values():
+            for cell in universe.cells.values():
+                if cell.fill_type in ('material', 'distribmat', 'void'):
+                    if universe not in material_universes:
+                        material_universes[universe.id] = universe
+
+        return material_universes
+
+    def get_all_lattices(self):
+        """Return all lattices defined
+
+        Returns
+        -------
+        collections.OrderedDict
+            Dictionary mapping lattice IDs to :class:`openmc.Lattice` instances
+
+        """
+        lattices = OrderedDict()
+
+        for cell in self.get_all_cells().values():
+            if cell.fill_type == 'lattice':
+                if cell.fill.id not in lattices:
+                    lattices[cell.fill.id] = cell.fill
+
+        return lattices
+
+    def get_all_surfaces(self):
+        """
+        Return all surfaces used in the geometry
+
+        Returns
+        -------
+        collections.OrderedDict
+            Dictionary mapping surface IDs to :class:`openmc.Surface` instances
+
+        """
+        surfaces = OrderedDict()
+
+        for cell in self.get_all_cells().values():
+            surfaces = cell.region.get_surfaces(surfaces)
+        return surfaces
+
+    def get_materials_by_name(self, name, case_sensitive=False, matching=False):
+        """Return a list of materials with matching names.
+
+        Parameters
+        ----------
+        name : str
+            The name to match
+        case_sensitive : bool
+            Whether to distinguish upper and lower case letters in each
+            material's name (default is False)
+        matching : bool
+            Whether the names must match completely (default is False)
+
+        Returns
+        -------
+        list of openmc.Material
+            Materials matching the queried name
+
+        """
+
+        if not case_sensitive:
+            name = name.lower()
+
+        all_materials = self.get_all_materials().values()
+        materials = set()
+
+        for material in all_materials:
+            material_name = material.name
```

Comment:
This can be simplified to `material_name = material.name if case_sensitive else material.name.lower()`